File: Medium/Spirally traversing a matrix/spirally-traversing-a-matrix.cpp
//{ Driver Code Starts
#include <bits/stdc++.h> 
using namespace std; 

// } Driver Code Ends
class Solution
{   
    public: 
    //Function to return a list of integers denoting spiral traversal of matrix.
    vector<int> spirallyTraverse(vector<vector<int> > matrix, int r, int c) 
    {
        int count = 0;
        int total = r*c;
        vector<int> ans;
        
        int startRow = 0;
        int startCol = 0;
        int endRow = r-1;
        int endCol = c-1;
        
        while(count<total){
            for(int i=startCol; count<total && i<=endCol; i++){
                ans.push_back(matrix[startRow][i]);
                count++;
            }
            startRow++;

            for(int i=startRow; count<total && i<=endRow; i++){
                ans.push_back(matrix[i][endCol]);
                count++;
            }
            endCol--;
            
            for(int i=endCol; count<total && i>=startCol; i--){
                ans.push_back(matrix[endRow][i]);
                count++;
            }
            endRow--;

            for(int i=endRow; count<total && i>=startRow; i--){
                ans.push_back(matrix[i][startCol]);
                count++;
            }
            startCol++;
        }
        return ans;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin>>t;
    
    while(t--) 
    {
        int r,c;
        cin>>r>>c;
        vector<vector<int> > matrix(r); 

        for(int i=0; i<r; i++)
        {
            matrix[i].assign(c, 0);
            for( int j=0; j<c; j++)
            {
                cin>>matrix[i][j];
            }
        }

        Solution ob;
        vector<int> result = ob.spirallyTraverse(matrix, r, c);
        for (int i = 0; i < result.size(); ++i)
                cout<<result[i]<<" ";
        cout<<endl;
    }
    return 0;
}
// } Driver Code Ends

File: Medium/Spirally traversing a matrix/spirally-traversing-a-matrix.py
#User function Template for python3

class Solution:
    def spirallyTraverse(self, matrix, r, c):
        if r < 1 or c < 1:
            return []
        loci = [(0, j) for j in range(c)]
        loci += [(i, c - 1) for i in range(1, r)]
        loci += [(r - 1, j) for j in range(c - 2, -1, -1) if r != 1]
        loci += [(i, 0) for i in range(r - 2, 0, -1) if c != 1]
        out = [matrix[i][j] for i, j in loci]
        for i in range(r):
            matrix[i] = matrix[i][1: c - 1]
        matrix = matrix[1: r - 1]
        return out + self.spirallyTraverse(matrix, r - 2, c - 2)


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        r,c = map(int, input().strip().split())
        values = list(map(int, input().strip().split()))
        k = 0
        matrix =[]
        for i in range(r):
            row=[]
            for j in range(c):
                row.append(values[k])
                k+=1
            matrix.append(row)
        obj = Solution()
        ans = obj.spirallyTraverse(matrix,r,c)
        for i in ans:
            print(i,end=" ")
        print()

# } Driver Code Ends